### PR TITLE
[Unity] Basic StructInfo Analysis and Expr construction

### PR DESCRIPTION
**Pull request by YuchenJin**
This PR adds struct info analysis and expr support. These are logics to construct the IR node and perform struct info related analysis.

Testcases are added to cover the IR node construction and related struct info analysis checks.

Co-Authored-by: Tianqi Chen <tianqi.tchen@gmail.com>
Co-Authored-by: Altan Haan <altanh@cs.washington.edu>
Co-Authored-by: Andrew Liu <andrewlliu@gmail.com>
Co-Authored-by: Hongyi Jin <3231950289@qq.com>
Co-Authored-by: Jiawei Liu <jaway.liu@gmail.com>
Co-Authored-by: Junru Shao <junrushao1994@gmail.com>
Co-Authored-by: Lesheng Jin <34279105+LeshengJin@users.noreply.github.com>
Co-Authored-by: masahi <masahi129@gmail.com>
Co-Authored-by: Prakalp Srivastava <prakalp@octoml.ai>
Co-Authored-by: Ruihang Lai <ruihangl@cs.cmu.edu>
Co-Authored-by: Siyuan Feng <Hzfengsy@sjtu.edu.cn>
Co-Authored-by: Steven S. <Lyubomirsky slyubomirsky@octoml.ai>
Co-Authored-by: Sunghyun Park <49998730+sunggg@users.noreply.github.com>
Co-Authored-by: Yixin Dong <ubospica@gmail.com>
Co-Authored-by: Yong Wu <yongcale@gmail.com>
Co-Authored-by: Ziheng Jiang <ziheng@apache.org>

### CODE REVIEW

**Review by Mousius — CHANGES_REQUESTED** [review]
Can you please add some test cases around all of this new logic?

File: include/tvm/ir/type.h
Comment (tqchen) [inline include/tvm/ir/type.h:135]: NOTE for followup: we can send this change to main

File: include/tvm/relax/utils.h
Comment (tqchen) [inline include/tvm/relax/utils.h:141]: likely we only need IsLeafOrTuple, and can skip Bind/IsBoolScalarType/CopyWithNewParams

File: include/tvm/relax/op_attr_types.h
Comment (tqchen) [inline include/tvm/relax/op_attr_types.h:43]: let us only keep FInferStructInfo and remove the rest for now

**Comment by tqchen** [comment]
Thank you @YuchenJin, left some high level comments.

File: include/tvm/ir/type.h
Comment (YuchenJin) [reply include/tvm/ir/type.h:136]: Agree, will do in a later PR.

**Comment by YuchenJin** [comment]
> Can you please add some test cases around all of this new logic?

Hi @Mousius, after discussing with @tqchen, we reorganized this PR and added test cases. Please take another look, thanks!

**Review by tqchen — APPROVED** [review]
Thank you for adding the testcases

**Review by junrushao — APPROVED** [review]
LGTM!

**Comment by Mousius** [comment]
@junrushao it is not inclusive to merge changes over the weekend without giving time to for those who have requested changes to re-review and it is not inclusive to expect any potential reviewers to have seen this change since it was raised and merged in just under 25 hours at a time when most people aren't active. Please could you be more considerate  of the wider community in future.

**Comment by junrushao** [comment]
Thank you for your comment. 

The flexibility of branch development offers opportunities for us to experiment with possible modes of agile development with quality. Agility in the scaffolding phase would help us to migrate development sooner to empower the community members who would like to have agile development and enable quicker migration as in [branch establishment guideline](https://github.com/apache/tvm/issues/13876). So I merged the code after I read through the current review state and ensured existing review requests on test cases are addressed adequately.

I also share the same goal of incorporating a broad set of suggestions and comments from the wider community as part of the development process and would love to experiment ways to enable that together in the agile branch development.

Specifically, we love follow-up comments and suggestions in this PR thread or forum threads for topics related to this PR (and not close such opportunities at the merge point). We would love to see additional valuable comments that can be made as follow-up requests in the PR thread, forum, or community meetings. As someone who owns the maintenance of related areas, I will do my best to address them in a continuous fashion.

As the merger I am committed to such follow up-suggestions, requests, and bug fixes caused by the merge. Thank you everyone for valuable review comments. I will love to have continued conversations to improve unity in general.

**Comment by Mousius** [comment]
Thanks for your reply @junrushao,

> The flexibility of branch development offers opportunities for us to experiment with possible modes of agile development with quality. Agility in the scaffolding phase would help us to migrate development sooner to empower the community members who would like to have agile development and enable quicker migration as in [branch establishment guideline](https://github.com/apache/tvm/issues/13876). So I merged the code after I read through the current review state and ensured existing review requests on test cases are addressed adequately.

Agile development is not justification to circumvent the review process established within TVM if the aim is to eventually merge `unity` into `main`; if this is the case, `unity` should be held to the same review guidelines as well as respecting the existing committers in the community. Agile development also prioritises working software, which benefits from high quality code reviews - given that both yourself and @tqchen are listed as co-authors but test coverage was missing in this PR, it would be good to get more eyes on the patches before we create a foundation of technical debt in the project which eventually gets merged into `main`.

> I also share the same goal of incorporating a broad set of suggestions and comments from the wider community as part of the development process and would love to experiment ways to enable that together in the agile branch development.

As an open source project, there needs to be a better balance between merging things quickly and allowing people visibility of changes - minimally this is at least leaving a large PR such as this open for a few days before merging (as specified in our [code review guidelines](https://github.com/apache/tvm/blob/main/docs/contribute/code_review.rst)) to allow others to comment and being respectful of those who do take the time to review.

> Specifically, we love follow-up comments and suggestions in this PR thread or forum threads for topics related to this PR (and not close such opportunities at the merge point). We would love to see additional valuable comments that can be made as follow-up requests in the PR thread, forum, or community meetings. As someone who owns the maintenance of related areas, I will do my best to address them in a continuous fashion.

As I mentioned above, should you wish to merge `unity` into `main`, I would suggest following practices established within the project ([code review guidelines](https://github.com/apache/tvm/blob/main/docs/contribute/code_review.rst)) rather than excluding the wider community, there's good advice on [community participation](https://github.com/apache/tvm/blob/main/docs/contribute/code_review.rst#community-participation) detailing a lot of the issues raised above.

**Comment by tqchen** [comment]
Thank you, everyone, for chiming in.

In order to get unity merged into main, we will need to ensure the end state (at the merging proposal) of the unity branch gets to a ready state. In the meantime, the branch development offers some flexibility for us to explore ways to gain agility with quality. We can have continuous conversations through the course of branch development to iterate the modules so we can answer the questions and continue to iterate on the foundation before merging the proposal.

There is no disagreement on goals, such as having high-quality solid foundations while collecting community inputs not only at the PR merging point but through continuous periods of time during development. 

For the branch establishment period specifically, some initial agility is helpful to get the initial scaffolds and enable quicker migration can benefit the community. We can also leverage the branch development opportunity to have continuous conversations instead of stopping at PR merges. For example, suggestions about changes, questions in this thread, or follow-up thread (aka, nothing is yet final at the merging point, and the community can continue to provide inputs and iterate on the solutions). To make sure we have accountability for quality and continuous community participation, the merger should also be responsible for addressing these follow-up inputs, suggestions, and questions and resolving foundational technical debts that the merger may have caused. This is a bit like a mixture of review-commit-review models. We can also continuously monitor the experiments and see, for example, if there are any lessons like technical debts that do arise, and adjust accordingly.

Coming specifically to the technical part of the current PR. Please do bring followup suggestions at any time point, in this thread or other places about the particular change. The initial suggestions about test coverage is very valuable and I see yuchen have updated the proposal set with proper test cases. Suggestions of possible additional test cases or possible technical debts are very welcomed. At the moment the test covers the struct info analysis and expr construction, hopefully more tests will be enabled after the parser is enabled through a rich set of TVMScript.


Thank you

**Comment by Mousius** [comment]
Hi @tqchen,

> In order to get unity merged into main, we will need to ensure the end state (at the merging proposal) of the unity branch gets to a ready state.

Merging large branches such as `unity` makes it incredibly hard to review in detail and the review process is there to help us review incremental changes. Incrementally reviewing is a standard agile practice which typically results in fewer defects overall ([great post on that here](https://smallbusinessprogramming.com/optimal-pull-request-size/)), even if it does take marginally longer for each individual change to land. I'm also happy to have a discussion about whether we should move changes to a further PR, as in our existing [code review guidelines](https://github.com/apache/tvm/blob/main/docs/contribute/code_review.rst) part of the process is communication and trust. 

> Suggestions of possible additional test cases or possible technical debts are very welcomed. 

By merging this outside of the already agreed process in [the code review guidelines](https://github.com/apache/tvm/blob/main/docs/contribute/code_review.rst) you've already excluded members of the community who contribute primarily through code review and primarily during working hours. The evidence is in this PR, had I not been active at the weekend this pull request would've been raised and merged without test coverage and would therefore have contributed to the technical debt in the `unity` branch.

**Comment by tqchen** [comment]
Thanks @Mousius, thank you for your inputs .  

Agree that it is a challenge to be able to get ramped up and look into the set of changes just in a PR. The good part of branch development is that we have a long extended period of time, where multiple evidences, feedbacks, engagements can be provided, to get the community ramped up, participate, and review the set of changes as a whole through a long extended period of time.

The idea is that even though things may not have been caught immediately by one of us(such as test coverage), having a quick followup comment such as (please add test coverage) will likely result in a quick PR from the merger to update the test case, as a result we quickly get to the state of readiness. In this case, likely I would also suggested the same, and as a result covers the test. So such iterative engagement and feedbacks helps us to continue improve the changes.

**Comment by tqchen** [comment]
I would like to followup and say that when we experiment different modes of operations at different period of the development. My take is that we will gain more parallelism after the branch establishment, and we can update our modes of operations at that time point. In the meantime, continuous suggestions, discussions and iterations will likely helps us to get to a better foundation even faster.

**Comment by Mousius** [comment]
> The idea is that even though things may not have been caught immediately by one of us(such as test coverage), having a quick followup comment such as (please add test coverage) will likely result in a quick PR from the merger to update the test case, as a result we quickly get to the state of readiness. In this case, likely I would also suggested the same, and as a result covers the test. So such iterative engagement and feedbacks helps us to continue improve the changes. The door is not closed at merging point.
> 
> There are also more ways that we can help to ramp up in the related code areas and bringing suggestions, participating in [community forums](https://discuss.tvm.apache.org/c/development/unity/14), bringing in suggestions are all valuable ways to help us continue to improve the set of changes over the branch.

Thanks for the clarification @tqchen, it is clear that the `unity` branch is not intended to be as inclusive of the wider community as the `main` branch and favours specific kinds of contributors; this PR and https://github.com/apache/tvm/pull/13926 are clear indicators of that, and it is deeply saddening that we have existing guidelines to help make TVM more inclusive yet the `unity` branch has chosen not to follow them for such a large contribution.

**Comment by tqchen** [comment]
Thanks, @Mousius as I clarified, the branch development in `unity` offers some flexibility to help us through the branch establishment stage. 

The reality is that we want to do more for our community over an extended period of time. For example, having continuous forum threads, and enable follow-up comments on particular changes or follow-up changes, and being ready to bring future iterations. That also includes examples and discussions on how certain modules can be improved at any time point of branch development (after merging). 

Looking at the state of the tvm project specifically. Few proposals would come with tutorials, demos, lectures, continued forum thread discussions, or meetup sessions that discussions the design perspectives. These are useful tools for us to reach a much wider community and will help us to get a better quality foundation through this process together. They also make our approach more inclusive to those who need a broader set of means to engage, provide feedback and contribute.

The branch development offers some flexibility and helps us to adapt our mode of development at different stages of the module. As we complete the branch establishment phase, we can update the approach accordingly, which will likely get closer to normal steady development.

Please feel free to open up [forum threads](https://discuss.tvm.apache.org/c/development/unity/14), provide suggestions in the comment block, ask questions, or provide feedback by any other means. That is the reason why we are developing in a branch so the feedbacks are not necessarily quickly come at a stall at merging point. We love all these ways of participations together in the community.

**Pull request merged**
